### PR TITLE
feat - Allow advanced merging through `mergeWithRules`

### LINKED ISSUE #146: Merging two module.rules just appends
This is the test I want to run but it seems that the latest version does not longer merge rules.

```ts
  it('Should merge loader options', () => {
    const output = merge(
      {
        module: {
          rules: [
            {
              test: /\.css$/,
              use: [{ loader: 'style-loader' }, { loader: 'sass-loader' }],
            },
          ],
        },
      },
      {
        module: {
          rules: [
            {
              test: /\.css$/,
              use: [
                {
                  loader: 'style-loader',
                  options: {
                    modules: true,
                  },
                },
              ],
            },
          ],
        },
      }
    );

    const expected = {
      module: {
        rules: [
          {
            test: /\.css$/,
            use: [
              {
                loader: 'style-loader',
                options: {
                  modules: true,
                },
              },
              { loader: 'sass-loader' },
            ],
          },
        ],
      },
    };

    expect(output).toEqual(expected);
  });
```

Output is:

```
  [
        { test: /\.css$/, use: [ [Object], [Object] ] },
        { test: /\.css$/, use: [ [Object] ] }
      ]
```

### PR DESCRIPTION
**Pull request by bebraw**
See the readme for the full example.

Closes #146 and #149.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -57,23 +53,136 @@ function mergeWithCustomize(options: ICustomizeOptions) {
...
+      if (v === CustomizeRule.Match) {
+        rulesToMatch.push(k);
+      } else {
+        operations[k] = v;
+      }
+    });
+
+    const bMatches = b.filter(o => {
+      const matches = rulesToMatch.every(
+        rule => ao[rule].toString() === o[rule].toString()
```
Comment (filoxo) [inline src/index.ts:127]: I tried testing this on a local project, and am getting this error: 

```
[app]/node_modules/webpack-merge/dist/index.js:122
            var matches = rulesToMatch.every(function (rule) { return ao[rule].toString() === o[rule].toString(); });
                                                                               ^

TypeError: Cannot read property 'toString' of undefined
    at [app]/node_modules/webpack-merge/dist/index.js:122:80
    at Array.every (<anonymous>)
    at [app]/node_modules/webpack-merge/dist/index.js:122:40
    at Array.filter (<anonymous>)
    at [app]/node_modules/webpack-merge/dist/index.js:121:26
    at Array.map (<anonymous>)
    at mergeWithRule ([app]/node_modules/webpack-merge/dist/index.js:108:17)
    at mergeWithRules ([app]/node_modules/webpack-merge/dist/index.js:101:16)
    at [app]/node_modules/webpack-merge/dist/index.js:90:16
    at _joinArrays ([app]/node_modules/webpack-merge/dist/join-arrays.js:44:50)
```

due to having a [`parser` config](https://webpack.js.org/configuration/module/#ruleparser) in my rules: 

```js
rules: [
  {
    parser: {
      system: false,
    },
  },
  {
    test: /\.(js|ts)x?$/,
    ...
  },
  {
    test: /\.css$/i,
    ...
  },
],
```

For testing, I removed that rule completely and then started getting the error below because of the `exclude: /node_modules/` in the next rule.

```
[app]/node_modules/webpack-merge/dist/index.js:110
    var ret = a.map(function (ao) {
                ^

TypeError: a.map is not a function
    at mergeWithRule ([app]/node_modules/webpack-merge/dist/index.js:110:17)
    at [app]/node_modules/webpack-merge/dist/index.js:161:30
    at Array.forEach (<anonymous>)
    at [app]/node_modules/webpack-merge/dist/index.js:130:28
    at Array.map (<anonymous>)
    at mergeWithRule ([app]/node_modules/webpack-merge/dist/index.js:110:17)
    at mergeWithRules ([app]/node_modules/webpack-merge/dist/index.js:102:16)
    at [app]/node_modules/webpack-merge/dist/index.js:90:16
    at _joinArrays ([app]/node_modules/webpack-merge/dist/join-arrays.js:44:50)
    at [app]/node_modules/webpack-merge/dist/merge-with.js:32:17
```


I imagine edge cases like this are the reason you've redone the API so this might not be something you want to handle. Just figured I might report it early in case it helps.

Comment (bebraw) [reply src/index.ts:127]: Thanks for testing! Can you send me the exact merge you tried so I can convert that as a test case to support?

> I imagine edge cases like this are the reason you've redone the API so this might not be something you want to handle. Just figured I might report it early in case it helps.

Exactly. With this PR we'll get a lot of the power of the old API back while being explicit about it.

Comment (filoxo) [reply src/index.ts:127]: Sure thing. I used the [syntax in the README](https://github.com/survivejs/webpack-merge/pull/150/files#diff-04c6e90faac2675aa89e2176d2eec7d8R197-R209), and the relevant parts of the configs are below.

<details>
    <summary>Base config</summary>

```js
rules: [
  {
    parser: {
      system: false,
    },
  },
  {
    test: /\.(js|ts)x?$/,
    exclude: /node_modules/,
    use: {
      loader: require.resolve("babel-loader", { paths: [__dirname] }),
    },
  },
  {
    test: /\.css$/i,
    include: [/node_modules/, /src/],
    use: [
      {
        loader: require.resolve("style-loader", { paths: [__dirname] }),
      },
      {
        loader: require.resolve("css-loader", { paths: [__dirname] }),
        options: {
          modules: false,
        },
      },
    ],
  },
],
```

</details>

<details>
    <summary>Consumer config</summary>

```js
rules: [
  {
    test: /\.html$/i,
    use: ["html-loader"],
  },
  {
    test: /\.css$/i,
    use: [
      require.resolve("style-loader", { paths: [__dirname] }),
      {
        loader: require.resolve("css-loader", { paths: [__dirname] }),
        options: {
          importLoaders: 1,
        },
      },
      {
        loader: "postcss-loader",
      },
    ],
  },
],
```

</details>

Comment (bebraw) [reply src/index.ts:127]: Can you still paste the merge you tried?

Comment (filoxo) [reply src/index.ts:127]: Is this what you're asking? I'm sorry if I'm misunderstanding. 

```js
const defaultConfig = {
    module: {
      rules: [
        {
          parser: {
            system: false,
          },
        },
        {
          test: /\.(js|ts)x?$/,
          exclude: /node_modules/,
          use: {
            loader: require.resolve("babel-loader", { paths: [__dirname] }),
          },
        },
        {
          test: /\.css$/i,
          include: [/node_modules/, /src/],
          use: [
            {
              loader: require.resolve("style-loader", { paths: [__dirname] }),
            },
            {
              loader: require.resolve("css-loader", { paths: [__dirname] }),
              options: {
                modules: false,
              },
            },
          ],
        },
      ],
    },
  }

  const localConfig = {
    module: {
      rules: [
        {
          test: /\.html$/i,
          use: ["html-loader"],
        },
        {
          test: /\.css$/i,
          use: [
            {
              loader: require.resolve("style-loader", { paths: [__dirname] }),
            },
            {
              loader: require.resolve("css-loader", { paths: [__dirname] }),
              options: {
                importLoaders: 1,
              },
            },
            {
              loader: "postcss-loader",
            },
          ],
        },
      ],
    }
  }

  const config = mergeWithCustomize({
    customizeArray: customizeArray({
      module: {
        rules: {
          test: "match",
          use: {
            loader: "match",
            options: "replace"
          }
        }
      }
    })
  })(defaultConfig, localConfig)
```

I also tried with `use: "replace"` in the `customizeArray` object since in my case I wouldn't mind completely overriding all the other loaders if the test regexes match, but got the same error.

Comment (bebraw) [reply src/index.ts:127]: That's perfect. Thanks!

Comment (bebraw) [reply src/index.ts]: @filoxo I pushed a fix for this particular case. Please give it a go.

Note that for the system to work as you expect, you should use the form `use: [{ loader: "html-loader" }]` for each loader. See the test I added for the complete solution.

Comment (filoxo) [reply src/index.ts]: Just tested it out and its working like a charm! I think consistency in the loader syntax makes total sense. I did notice that in the merged object the `style-loader` rule now has an `options` key (which is `undefined` so it won't cause any issues as-is) while neither source object had options (see below). I imagine that's an artifact of the comparison logic. 

```js
{
  module: {
    rules: [
      { parser: { system: false } },
      {
        test: /\.(js|ts)x?$/,
        exclude: /node_modules/,
        use: {
          loader: '[app]/node_modules/babel-loader/lib/index.js'
        }
      },
      {
        test: /\.css$/i,
        include: [ /node_modules/, /src/ ],
        use: [
          {
            loader: '[app]/node_modules/style-loader/dist/cjs.js',
            options: undefined // TODO: this is weird, neither source object had an options value yet its key exists now
          },
          {
            loader: '[app]/node_modules/css-loader/dist/cjs.js',
            options: { importLoaders: 1 }
          },
          { loader: 'postcss-loader' }
        ]
      },
      { test: /\.html$/i, use: [ { loader: 'html-loader' } ] }
    ]
  },
}
```

I really appreciate the quick fix! Hopefully others benefit from this as well.

Comment (bebraw) [reply src/index.ts]: Yeah, the options property is something I could likely remove from the result as well.

Comment (bebraw) [reply src/index.ts]: @filoxo I fixed the `undefined` issue and also simplified the API a notch. Please have a look. Once you ok the feature, I'll publish. 👍

Comment (filoxo) [reply src/index.ts]: Thanks for the update @bebraw. I tried out the new code and it seems to have reverted in behavior (despite that it passes in your local tests). I'm sort of stumped as to why that would be since my code hasn't changed. I threw together this example repo to show what I'm seeing since its a little more complex: https://github.com/filoxo/webpack-merge-test The instructions are included on how to run, and what should be expected. Let me know if you'd prefer something else to triage this.

Comment (bebraw) [reply src/index.ts]: @filoxo I moved the behavior behind a new function, `mergeWithRules`. I see your code isn't using that yet.

Comment (bebraw) [reply src/index.ts]: @filoxo Here's usage: https://github.com/survivejs/webpack-merge/pull/150/files#diff-04c6e90faac2675aa89e2176d2eec7d8R227 .

Comment (filoxo) [reply src/index.ts]: 🤦  I totally missed that, sorry! Just switched to `mergeWithRules` and it looks like its merged correctly now.

Comment (bebraw) [reply src/index.ts]: No probs. I think I'll merge this tomorrow then and publish a new minor version with the feature. 👍
